Add support for bootstrap containers via API settings

arnaldo2792 opened the pull request:
**Issue number:**
bottlerocket-os/bottlerocket#1392

**Description of changes:**

```
5153d08c host-containers: change permissions in `/etc/host-containers`
```
This commit updates the permissions in `/etc/host-containers` from `755` to `750`

```
d59628c7 models: change ContainerImage to HostContainer
```
This commit updates the model name used to configure a host container from `ContainerImage` to `HostContainer`

```
8ce2cae3 Add support for bootstrap-containers via settings
```
This commit adds support to create bootstrap containers through the API. Bootstrap containers are host containers that can be used to setup the host during the execution of the `configured` target.

These containers are created with the prefix `boot` to prevent collisions with normal host containers. They can be setup to fail the boot process if the underlying container task exists with a non-zero status code.

```
f5e29ce9 host-ctr: support for bootstrap containers
```
This commit adds support for bootstrap containers in `host-ctr`. The  `container-type` flag was added to specify the type of container to be setup, which defaults to `host`. This allows to add support for other type of containers in the future without adding flags for each of them.

The root filesystem mounts were refactored out from `withSuperpowered` to `withRootFilesystemMounts`, since bootstrap containers require access to the underlying root filesystem without being given the superpowered treatment.

Failed container tasks will now return an error if the task exited with a non-zero status.

**Testing done:**
In `k8s 1.18`, `ecs` and `dev` aarch64:

* Run `systemctl status` to check that no units were failing
* Create `nginx` pods/containers, access them and `curl http://localhost` successfully
* Verifiy that both the control and admin host containers are running, with the proper configurations
* Execute `systemctl isolate multi-user` to verify that no services were killed/restarted
* Migrate from 1.0.7 to 1.0.8, back and forth to verify the new configurations were deleted/created
* Create a bootstrap container with the following configurations:
```toml
# user-data
[settings.bootstrap-containers.admin]
mode = "once"
source = "docker.io/arnaldo2792/sleeper:latest"
user-data = "ypXCt82h4bSlwrfKlA==" # base64 of ʕ·͡ᴥ·ʔ
``` 
```Dockerfile
# Image definition
FROM alpine
ADD bootstrap-script /
RUN chmod +x /bootstrap-script
ENTRYPOINT ["sh", "bootstrap-script"]
```

```sh
# bootstrap-script
set -e
CURRENT_DIR=/.bottlerocket/bootstrap-containers/current
BEAR_DIR=/.bottlerocket/bootstrap-containers/admin
VAR_DIR=/.bottlerocket/rootfs/var
MY_DIR=$VAR_DIR/lib/my_directory
cat $BEAR_DIR/user-data
touch $CURRENT_DIR/access
mkdir $MY_DIR
chmod -R o+r $MY_DIR
chown -R 1000:1000 $MY_DIR
sleep 120
```
* Verified that:
    * `ʕ·͡ᴥ·ʔ` appears in the journal logs
    * The directory `/var/lib/my_directory` was created, with the expected user/group ids, and permissions
    * The file `/local/bootstrap-containers/access` created through the `current` mount
    * `configured.target` was "blocked" for two minutes
    * `mark-successful-boot.service` executed successfully
    * After a reboot, the bootstrap container wasn't executed

* Update the bootstrap container's configuration with `apiclient set bootstrap-containers.admin.mode=once`, reboot and verified its execution failed because `/var/lib/my_dir` already exists.

```sh
Starting bootstrap container admin...
[8.350595] host-ctr[456]: mkdir: can't create directory '/.bottlerocket/rootfs/var/lib/my_directory': File exists
[ 8.415213] host-ctr[456]: time="2021-04-02T00:54:43Z" level=fatal msg="Container exited with non-zero status"
[FAILED] Failed to start bootstrap container admin.
```

* Update the bootstrap container's configuration with `apiclient set bootstrap-containers.admin.essential=true`, reboot and verified the boot failed:

```sh
[FAILED] Failed to start bootstrap container admin.
See 'systemctl status bootstrap-containers@admin.service' for details.
[DEPEND] Dependency failed for Bottlerocket final configuration complete.
[DEPEND] Dependency failed for Isolates multi-user.target.
[FAILED] Failed to start Isolates configured.target.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
host-ctr: support for bootstrap containers

This commit adds support for bootstrap containers in `host-ctr`. The
`container-type` flag was added to specify the type of container to be
setup, which defaults to `host`. This allows to add support for other
type of containers in the future without adding flags for each of them.

The root filesystem mounts were refactored out from `withSuperpowered`
to `withRootFilesystemMounts`, since bootstrap containers require access
to the underlying root filesystem without being given the superpowered
treatment.

Failed container tasks will now return an error if the task exited with
a non-zero status.

arnaldo2792 committed:
Add support for bootstrap-containers via settings

This commit adds support to create bootstrap containers through the
API. Bootstrap containers are host containers that can be used to setup
the host during the execution of the `configured` target.

These containers are created with the prefix `boot` to prevent
collisions with normal host containers. They can be setup to fail the
boot process if the underlying container task exists with a non-zero
status code.

bcressey commented on packages/os/host-containers@.service:
```diff
@@ -19,4 +19,4 @@ TimeoutStopSec=60
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target
```
Ordinary host containers should get pushed out into a later phase.

They should probably have `After` / `Wants` on `configured.target`, and continue to be `WantedBy` on `multi-user.target`.

arnaldo2792 replied:
Done and tested! Ordinary host containers now start on `configured.target`

bcressey commented on packages/host-ctr/host-containerd.service:
```diff
@@ -1,8 +1,9 @@
-After=network-online.target configured.target
-Wants=network-online.target configured.target
+After=network-online.target
+Before=preconfigured.target
+Wants=network-online.target preconfigured.target
```
The `Before` / `Wants` relationship here is strange.

We don't want to start this service until we've done all the initial configuration work, since some values like proxy server settings may not be ready yet.

I expect it's more correct to change `After` / `Wants` to `preconfigured.target`, and `WantedBy` to `configured.target`.

arnaldo2792 replied:
Done

bcressey commented on packages/os/bootstrap-containers@.service:
```diff
@@ -0,0 +1,22 @@
+Type=oneshot
+EnvironmentFile=/etc/bootstrap-containers/%i.env
+Environment=LOCAL_DIR=/local
+# Create directory for container persistent storage
```
We're missing a `mkdir` here.

arnaldo2792 replied:
Actually, that comment shouldn't be there. The persistent storage directory for the bootstrap container will be created by the `bootstrap-containers` binary, at the `preconfigured` stage. So when the bootstrap containers' systemd units are started on the `configured` target, the directories will be already there, that's why I omitted the `mkdir` as part of the unit.

bcressey replied:
Can we clean up `host-containers@.service` as well, then?

bcressey commented on packages/release/activate-configured.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=Isolate configured.target
+After=preconfigured.target host-containerd.service
```
`After` / `WantedBy` `preconfigured.target` is odd but seems right.

However the dependency on `host-containerd.service` does not seem correct; ideally we'd let that be started in phase two as part of `configured.target`.

arnaldo2792 replied:
Done

bcressey commented on packages/release/configured.target:
```diff
@@ -1,5 +1,5 @@
 Description=Bottlerocket user and dynamic configuration complete
-After=settings-applier.service
-Requires=settings-applier.service early-boot-config.service
+After=preconfigured.target
+Wants=preconfigured.target
```
```suggestion
After=preconfigured.target
Wants=preconfigured.target
WantedBy=multi-user.target
```

arnaldo2792 replied:
I addressed the changed, but `WantedBy` should be in the `[Install]`, section, which you can see underneath this comment.

bcressey commented on packages/os/settings-applier.service:
```diff
@@ -12,4 +12,4 @@ RemainAfterExit=true
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target
```
Previously, this unit was effectively `RequiredBy: configured.target`.
```suggestion
RequiredBy=preconfigured.target
```

arnaldo2792 replied:
Done

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -14,4 +14,4 @@ RemainAfterExit=true
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target
```
Previously, this unit was effectively `RequiredBy: configured.target`.

```suggestion
RequiredBy=preconfigured.target
```

arnaldo2792 replied:
Updated to `RequiredBy`

bcressey commented on packages/systemd/systemd.spec:
```diff
@@ -213,6 +213,9 @@ install -p -m 0644 %{S:3} %{buildroot}%{_cross_libdir}/systemd/journald.conf.d/j
 # with container networking by attempting to manage veth devices.
 rm -f %{buildroot}%{_cross_libdir}/systemd/network/*
 
+# Remove default target created by systemd
```
```suggestion
# Remove default target created by systemd, since we override it elsewhere.
```

arnaldo2792 replied:
Updated to mention that the default systemd unit is created in the release spec

bcressey reviewed (commented):
Nice!

I've reviewed the changes to the unit dependencies; haven't looked at the actual bootstrap containers implementation yet.

tjkirch commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -558,3 +558,44 @@ impl TryFrom<&str> for Lockdown {
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ValidContainerMode {
```
```suggestion
pub struct ContainerMode {
```
I regret the `Valid` prefix on `Base64`... I was trying to disambiguate from the library name it's wrapping, but any type we create should be "valid."  Inventions of our own, in particular, don't have any other state.  A container mode can be "off," "once," or "always," because that's how it's defined, and anything else isn't a container mode at all - no need to say valid.

arnaldo2792 replied:
Done

zmrow commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -154,7 +161,7 @@ func App() *cli.App {
 }
 
-func runCtr(containerdSocket string, namespace string, containerID string, source string, superpowered bool) error {
+func runCtr(containerdSocket string, namespace string, containerID string, source string, superpowered bool, bootstrap bool) error {
```
Since this function's definition is getting fairly large, would it make sense to use a single struct containing the arguments (essentially the "config" of the container to run)?  (The Go folks might have more to say about how idiomatic this would be).

samuelkarp replied:
Since this is a private function that really only gets called in one place, I don't know if we need to do that yet.  One change we could make is to separate this function a bit more.  The first two arguments here are only used up until line 181 for constructing the client; if we extract the client object from this function and instead pass it in (either directly or as a data field on a receiver object) that might be a bit cleaner.  I think that's the route I would go.

zmrow commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -206,6 +213,15 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
+		var persistentPath string
+
+		if bootstrap {
+			persistentPath = "bootstrap-containers/" + containerID
```
Do these make sense as constants?

arnaldo2792 replied:
I used a different approach now, let me know if you still think the new approach should use constants

zmrow commented on README.md:
```diff
@@ -449,6 +449,36 @@ We use it for the control container because it needs to be available early to gi
+
+Bootstrap containers have three different modes:
+
+* `always`: with this setting, the container is executed in all boots.
```
s/in all boots/on every boot

arnaldo2792 replied:
I like that better, thanks! :+1:

zmrow commented on README.md:
```diff
@@ -449,6 +449,36 @@ We use it for the control container because it needs to be available early to gi
+Bootstrap containers have three different modes:
+
+* `always`: with this setting, the container is executed in all boots.
+* `off`: blocks the execution of the container
```
It might be clearer to say "the container will not run"

arnaldo2792 replied:
:+1:

zmrow commented on sources/api/README.md:
```diff
@@ -123,12 +123,9 @@ Some day we may need to make an earlier service (say, apiserver) configurable th
-
-Applications can depend on this in their service definition.
-Services like Kubernetes and containerd depend on this.
+Bottlerocket uses a custom default systemd target called `preconfigured`, which is used to initialize all the services used by the API system (i. e. [settings-applier](#settings-applier)). Once the target is reached, a systemd unit isolates the `configured` target, which represents the point at which the system is fully configured. Applications can depend on this in their service definition. Serv ...
```
Our convention is typically to write markdown docs with one sentence per line.  If there aren't any blank lines between the sentences, they are formatted as a paragraph.  Doing it this way makes it a bit easier to read/edit.  For example:
```
this is a sentence.
this is another sentence.

this will be a new paragraph.
with another sentence.
```

webern commented on sources/api/bootstrap-containers/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+snafu = "0.6"
+tokio = { version = "0.2", default-features = false, features = ["macros", "rt-threaded"] }
+# When hyper updates to tokio 0.3:
+#tokio = { version = "0.3", default-features = false, features = ["macros", "rt-multi-thread"] }
```
```suggestion
#tokio = { version = "1", default-features = false, features = ["macros", "rt-multi-thread"] }
```

arnaldo2792 replied:
Updated

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,567 @@
+}
+
+/// Parse the args to the program and return an Args struct
+fn parse_args(args: env::Args) -> Args {
```
Can you change the arg parsing pattern to what we have in `shibaken` and `static-pods`? The reason is that the 'old' pattern of using `std::process:exit` makes argument parsing inherently un-testable. The new pattern (using a result) is testable.

```suggestion
fn parse_args(args: env::Args) -> Result<Args> {
```

arnaldo2792 replied:
I changed it to return `Result<Stuff>`

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,567 @@
+}
+
+/// Handles the `mark-bootstrap` subcommand
+async fn mark_bootstrap(args: Args) -> Result<()> {
```
```suggestion
async fn mark_bootstrap(args: MarkBootstrapArgs) -> Result<()> {
```

I think it would be better if all of the argument parsing were done in dedicated argument parsing functions.

The pattern to look at would be `apiclient`. You would have an enum like this:

```rust
enum Subcommand {
    MarkBootstrap(MarkBootstrapArgs),
    CreateContainers(CreateContainersArgs),
}
```

arnaldo2792 replied:
I updated the mechanism that I used to get command line arguments, to be similar to your proposal.

diranged commented on README.md:
```diff
@@ -449,6 +449,36 @@ We use it for the control container because it needs to be available early to gi
 Be careful, and make sure you have a similar low-level use case before reaching for host containers.
 
+#### Bootstrap containers settings
+* `settings.bootstrap-containers.<name>.source`: The URI of the container
```
Just curious - is this a single container we can run, or can we set up several? The plural on `bootstrap-containers` is confusing me a bit.

arnaldo2792 replied:
Hi @diranged, thanks for the feedback, I'll try to make this bit clearer. To answer your question, you could setup several bootstrap containers, but they should be independent from each other since we don't implement a mechanism to run them in a specific order.

diranged replied:
Ok great. Also, is there support for these containers being hosted in a private registry?

arnaldo2792 replied:
Since these host containers, you can use public/private ECR registries, as well as public registries.

diranged commented on README.md:
```diff
@@ -449,6 +449,36 @@ We use it for the control container because it needs to be available early to gi
+
+Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ecs-agent, kubelet and docker start.
+
+This type of containers are very similar to normal host containers, they come with persistent storage and with optional user data. Unlike normal host containers, bootstrap containers can't be threated as `superpowered` containers. However, these containers run as privileged and have access to the underlying root filesystem on `/.bottlerocket/rootfs`.
```
s/threated/treated/

arnaldo2792 replied:
Nice catch! Thanks

diranged commented on sources/api/bootstrap-containers/README.md:
```diff
@@ -0,0 +1,61 @@
+#!/bin/sh
+BEAR_DIR=/.bottlerocket/bootstrap-containers/bear
+cat $BEAR_DIR/user-data
+sleep 120
```
In this example, will boot be delayed by 120s until this completes?

arnaldo2792 replied:
Yes, that's correct. With this bootstrap container, systemd will "wait" for the container to exit. Upon completion, the boot process will continue.

diranged replied:
What happens if the container exits with a failure code?

arnaldo2792 replied:
Hi @diranged sorry for the late response. We added the flag `essential`, it can be set to `true` to fail the boot if the underlying bootstrap container exited with non-zero.

diranged commented on sources/api/bootstrap-containers/README.md:
````diff
@@ -0,0 +1,61 @@
+Given a bootstrap container with the following configuration:
+
+```toml
+[settings.bootstrap-containers.bear]
````
It looks like you can identify multiple bootstrap containers (negating my other question above). How do you order them - or are they explicitly unordered? That should just be called out in the docs..

arnaldo2792 replied:
I updated the documentation to mention that bootstrap containers are not executed in any order :+1: . Let me know if you have further questions.

webern commented on README.md:
```diff
@@ -458,6 +458,41 @@ We use it for the control container because it needs to be available early to gi
+
+Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ecs-agent, kubelet and docker start.
+
+This type of containers are very similar to normal host containers, they come with persistent storage and with optional user data.
```
nit

```suggestion
These types of containers are very similar to normal host containers, they come with persistent storage and with optional user data.
```

or

```
This type of container is very similar
```

or

```
Bootstrap containers are very similar
```

webern commented on README.md:
```diff
@@ -458,6 +458,41 @@ We use it for the control container because it needs to be available early to gi
+This type of containers are very similar to normal host containers, they come with persistent storage and with optional user data.
+Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
+However, these containers run as privileged and have access to the underlying root filesystem on `/.bottlerocket/rootfs`.
+Bootstrap containers are started once the `configured` systemd's target is reached, so they are not executed in a deterministic order.
```
I wonder if deterministic order would be 'nice-to-have' feature. One thing that's not clear currently in the documentation is whether the bootstrap containers are fired off in parallel or one-after-another.

arnaldo2792 replied:
I updated the documentation, and I mentioned that bootstrap containers run in parallel. Regarding the deterministic order, I think use cases that require specific order should belong to one container instead of splitting work and orchestrating containers. Do you have an specific use case in mind?

webern commented on sources/api/README.md:
```diff
@@ -123,12 +123,13 @@ Some day we may need to make an earlier service (say, apiserver) configurable th
-This is a systemd target that depends on [settings-applier](#settings-applier) and represents the point at which the system is fully configured.
+### Default systemd target and init workflow
 
+Bottlerocket uses a custom default systemd target called `preconfigured`, which is used to initialize all the services used by the API system (i. e. [settings-applier](#settings-applier)).
```
```suggestion
Bottlerocket uses a custom default systemd target called `preconfigured`, which is used to initialize all the services used by the API system (i.e. [settings-applier](#settings-applier)).
```

arnaldo2792 replied:
I removed this part of the documentation

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+}
+
+/// Parses user arguments into an Args struct
+fn parse_args(args: env::Args) -> Result<(Args, Subcommand)> {
```
I would expect the Subcommand to be a member of the args struct.

```suggestion
fn parse_args(args: env::Args) -> Result<(Args)> {
```

Also after seeing all of this arg parsing (it looks good), I wish I had recommended that we use `StructOpt`. But we don't have to do that now.

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+    if let Some(user_data) = &container_details.user_data {
+        debug!("Decoding user data for container '{}'", name);
+        let decoded_bytes = base64::decode(user_data.as_bytes()).context(error::Base64Decode {
+            base64_string: user_data.as_ref(),
```
Seems like that user_data could be quite large, do we want all of it in the error message?

arnaldo2792 replied:
This is what we do in [host-containers](https://github.com/bottlerocket-os/bottlerocket/blob/e04cdf50b21661e0e1633420f4b6381c920a71cc/sources/api/host-containers/src/main.rs#L358), do you have a different  approach in mind?

webern replied:
OK, seems a little untidy but not serious (unless it were seriously huge). An idea would be to give all of it if less than N bytes or MAX_BYTES if it is larger. Not worth blocking the feature for though!

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+
+        // Makes sure that the persistent storage directory exists
+        let dir = Path::new(PERSISTENT_STORAGE_DIR).join(name);
+        fs::create_dir_all(&dir).context(error::Mkdir { dir: &dir })?;
```
I don't know the answer: is there a potential race if multiple bootstrap containers are trying to create it? Presumably this function doesn't have an issue between time of checking for existence and when it tries to create the dir.

arnaldo2792 replied:
I don't think race conditions will occur while creating the directory. `PERSISTENT_STORAGE_DIR` will be created by the `tmpfilesd` configuration files before the bootstrap containers start. And the path `PERSISTENT_STORAGE_DIR/{name}` directory should be unique, because there can't be two bootstrap containers with the same name, since the `bootstrap-containers` settings is a Map.

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+            command(HOST_CTR_BIN, &["clean-up", "--container-id", name])?;
+        }
+    } else {
+        info!("Bootstrap container mode for '{}' is '{}'", name, mode);
```
Apologies but I don't understand how 'once' mode works.

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+
+    ensure!(
+        output.status.success(),
+        error::CommandFailure { bin_path, output }
```
Maybe add the exit code?

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+            KeyType::Data,
+            format!("settings.bootstrap-containers.{}.mode", container_id),
+        )
+        .expect("Adding container name to key resulted in invalid format");
```
I think you forgot to use `context` and an error type here.

webern commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,661 @@
+            source: model::modeled_types::error::Error,
+        },
+
+        #[snafu(display("'{}' failed - stderr: {}",
+                        bin_path, std::str::from_utf8(&output.stderr).unwrap_or_else(|_| "<invalid UTF-8>")))]
```
Probably better to display as best as we can rather than hide all of it.

```suggestion
        #[snafu(display("'{}' failed - stderr: {}", bin_path, output.stderr.to_string_lossy())]
```

arnaldo2792 replied:
Updated the code with `String::from_utf8_lossy`

webern commented on sources/api/migration/migrations/v1.0.8/add-bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,23 @@
+
+/// We added the setting `bootstrap-containers`
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
```
I don't think it matters, but just calling out that we are leaving any bootstrap container directories and env/userdata files behind on downgrade. That's probably fine and seems consistent with other new features we've added.

webern commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -206,6 +213,15 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 	// If the container doesn't already exist, create it
 	if container == nil {
+		// Set the destination name for the container persistent storage location
+		var persistentPath string
```
Nit: I think this is a directory name, not a complete path.

```suggestion
		var persistentDir string
```

arnaldo2792 replied:
I refactored the code a bit, and I used your suggestion. Let me know if you have any concern with the new approach I used.

webern commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -45,6 +45,9 @@ pub mod error {
         #[snafu(display("Invalid Kubernetes authentication mode '{}'", input))]
         InvalidAuthenticationMode { input: String },
 
+        #[snafu(display("Invalid bootstrap container mode '{}'", input))]
```
Nit: I think your error message in the bootstrap-containers binary will basically repeat this statement twice.

arnaldo2792 replied:
Updated the code to show the message just once

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -206,6 +213,15 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 	// If the container doesn't already exist, create it
 	if container == nil {
+		// Set the destination name for the container persistent storage location
+		var persistentPath string
+
+		if bootstrap {
+			persistentPath = "bootstrap-containers/" + containerID
+		} else {
+			persistentPath = "host-containers/" + containerID
+		}
```
Since you're treating non-boostrap as the default, it'd be slightly clearer to write the code like this:

```go
persistentDir := "host-containers/" + containerID
if bootstrap {
	persistentDir = "bootstrap-containers/" + containerID
}
```

arnaldo2792 replied:
I refactored the code, let me know if you have any concerns with the new approach I used

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -230,13 +246,15 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 					Options:     []string{"rbind", "rw"},
-					Destination: "/.bottlerocket/host-containers/" + containerID,
-					Source:      "/local/host-containers/" + containerID,
+					Destination: "/.bottlerocket/" + persistentPath,
+					Source:      "/local/" + persistentPath,
```
Is there a particular reason bootstrap containers get a different on-disk location than regular host containers?

arnaldo2792 replied:
I think the main reason is to prevent collisions if two containers have the same name

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -562,10 +580,24 @@ func withSuperpowered(superpowered bool) oci.SpecOpts {
+}
+
+// withPrivilegedAccess adds container options to mount the root filesystem
+// as well as privileged threatment
```
s/threatment/treatment/

arnaldo2792 replied:
I removed the `withPrivilegedAccess`, since bootstrap containers shouldn't have all privileges just yet.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -230,13 +246,15 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 				}}),
 			// Mount the rootfs with an SELinux label that makes it writable
 			withMountLabel("system_u:object_r:state_t:s0"),
 			// Include conditional options for superpowered containers.
 			withSuperpowered(superpowered),
+			// Include conditional options for bootstrap or superpowered containers
+			withPrivilegedAccess(superpowered || bootstrap),
```
Will `withSuperpowered` ever be called without `withPrivilegedAccess`?  The refactoring here moves some of the critical pieces of superpowered out into "privileged" that if you called `withSuperpowered` alone you wouldn't end up with a functional superpowered container.

arnaldo2792 replied:
I removed the `withPrivilegedAccess` method since bootstrap containers shouldn't have all privileges just yet.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -113,9 +114,15 @@ func App() *cli.App {
 					Value:       false,
 				},
+				&cli.BoolFlag{
+					Name:        "bootstrap",
```
Are "superpowered" and "bootstrap" mutually-exclusive, or can you have a superpowered bootstrap container?

arnaldo2792 replied:
Bootstrap containers can't be given the superpowered treatment (just yet), I added a validation to fail if the `superpowered` flag was passed and the container type is `bootstrap`.

samuelkarp commented on README.md:
```diff
@@ -458,6 +458,41 @@ We use it for the control container because it needs to be available early to gi
+* `settings.bootstrap-containers.<name>.mode`: The mode of the container, it could be one of `off`, `once` or `always`
+* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data
+
+Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ecs-agent, kubelet and docker start.
```
nit: "ecs-agent, kubelet, and docker start" to use the Oxford comma.

second nit: those are the names of the services in systemd, which is one correct way to do this.  The other way would be to use the names of the software which would be "ECS Agent", "Kubelet", and "Docker".  Is it your intention to use the systemd service names?

arnaldo2792 replied:
No, my intention was to used the services names. I already updated them.

samuelkarp commented on README.md:
```diff
@@ -458,6 +458,41 @@ We use it for the control container because it needs to be available early to gi
+Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
+However, these containers run as privileged and have access to the underlying root filesystem on `/.bottlerocket/rootfs`.
+Bootstrap containers are started once the `configured` systemd's target is reached, so they are not executed in a deterministic order.
+The boot process will "wait" for as long as the bootstrap containers run, and if any of them fail, the boot process will stop.
```
Might be worth adding a second sentence here: "Bootstrap containers must exit with an exit code of 0 before the boot process can continue."

arnaldo2792 replied:
Done :+1:

samuelkarp commented on README.md:
````diff
@@ -458,6 +458,41 @@ We use it for the control container because it needs to be available early to gi
+Here's an example of adding a bootstrap container with API calls:
+
+```
+apiclient -u /settings -X PATCH -d '{"bootstrap-containers": {"bootstrap": {"source": "MY-CONTAINER-URI", "mode": "once"}}}'
+apiclient -u /tx/commit_and_apply -X POST
````
Can this example be written with `apiclient set`?  Or are multiple fields required here such that the simplified `set` mode is insufficient?

arnaldo2792 replied:
I can write it with  `apiclient set --json` and specify the settings as a json blob or write multiple lines with `apiclient set bootstrap-containers.blah.*`

arnaldo2792 replied:
Done :+1:

samuelkarp commented on packages/os/bootstrap-containers@.service:
```diff
@@ -0,0 +1,21 @@
+EnvironmentFile=/etc/bootstrap-containers/%i.env
+Environment=LOCAL_DIR=/local
+ExecStart=/usr/bin/host-ctr run \
+    --container-id='%i' \
```
Are bootstrap containers allowed to have the same name as host containers?  If you're using the name as the container ID and they are allowed to have the same name, you won't actually be able to start both (host-ctr will see that a container already exists with the given ID and not create a new one).

arnaldo2792 replied:
I updated the bootstrap-containers initialization and they now are created in the `bootstrap` namespace to prevent collisions.

arnaldo2792 replied:
I changed the implementation to use prefixes instead of namespaces

samuelkarp reviewed (commented):
I haven't gotten through all the rust code in the third commit yet, but figured I'd stop here to get the questions I already have answered first.  Two things I'd like to get worked out:

1. If bootstrap containers can't be superpowered, we should enforce this assertion in the code of host-ctr
2. If host containers and bootstrap containers are allowed to share the same name, we need to work out non-collision for the container IDs in containerd

bcressey commented on packages/host-ctr/host-containerd.service:
```diff
@@ -19,4 +19,4 @@ LimitNOFILE=infinity
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target configured.target
```
I don't think this should be `WantedBy=preconfigured.target`.

bcressey commented on packages/os/host-containers@.service:
```diff
@@ -19,4 +19,4 @@ TimeoutStopSec=60
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=configured.target multi-user.target
```
These host containers should actually depend on `host-containerd.service` (via After + Wants) as well as `configured.target`.

They should not be `WantedBy=configured.target`, just `WantedBy=multi-user.target`.

bcressey commented on packages/release/activate-configured.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=Isolate configured.target
+After=preconfigured.target
```
We might want also want
```suggestion
After=preconfigured.target
Requires=preconfigured.target
```

bcressey commented on packages/release/configured.target:
```diff
@@ -1,5 +1,8 @@
-After=settings-applier.service
-Requires=settings-applier.service early-boot-config.service
+After=preconfigured.target
+Wants=preconfigured.target
```
For symmetry with `preconfigured.target`'s dependencies on `basic.target`:
```suggestion
Requires=preconfigured.target
```

bcressey commented on packages/release/activate-multi-user.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=Isolate multi-user.target
+After=configured.target
```
As with `activate-configured`:
```suggestion
Requires=configured.target
```

bcressey commented on packages/release/activate-multi-user.service:
```diff
@@ -0,0 +1,12 @@
+
+[Service]
+Type=oneshot
+ExecStart=systemctl isolate multi-user
```
As with `activate-configured`:
```suggestion
ExecStart=/usr/bin/systemctl isolate multi-user
```

bcressey commented on packages/release/activate-configured.service:
```diff
@@ -0,0 +1,12 @@
+
+[Service]
+Type=oneshot
+ExecStart=systemctl isolate configured
```
We should use the full path to `systemctl`, so we don't depend on any implicit `$PATH` setting:
```suggestion
ExecStart=/usr/bin/systemctl isolate configured
```

bcressey commented on packages/release/preconfigured.target:
```diff
@@ -0,0 +1,8 @@
+Description=Bottlerocket dynamic pre-configuration complete
+After=basic.target
+Requires=basic.target
+AllowIsolate=yes
```
We might also want:
```
RefuseManualStart=yes
RefuseManualStop=yes
```

For both targets, if it doesn't interfere with isolate.

bcressey commented on packages/systemd/systemd.spec:
```diff
@@ -212,6 +212,9 @@ install -p -m 0644 %{S:3} %{buildroot}%{_cross_libdir}/systemd/journald.conf.d/j
 # with container networking by attempting to manage veth devices.
 rm -f %{buildroot}%{_cross_libdir}/systemd/network/*
 
+# Remove default target created by systemd, since we override it in the release spec
+rm -f %{buildroot}%{_cross_libdir}/systemd/{system,user}/default.target
+
```
I'd also like to remove the stock `multi-user.target` from here, and add a copy to the release package that makes its dependencies explicit.

Then all of our targets will be packaged in the same place, and it will be somewhat easier to reason about the combined behavior.

arnaldo2792 commented:
### IMPORTANT UPDATE
I changed the base branch to have this and bottlerocket-os/bottlerocket#1423 reviewed in parallel

Changes in forced push:

* Replaced "fatals" with `Result<>` in bootstrap containers
* Use the containerd's namespace "bootstrap" for bootstrap containers to avoid collisions
* Removed the `withPrivilegedAccess` specs function, and replace it with `withRootFilesystemMounts`
* Added the `essential` flag on bootstrap containers, used to fail boots when set to `true`

arnaldo2792 commented:
Changes in forced push:

* Update the commits messages
* Update bootstrap-containers documentation
* Use the `BOOTSTRAP_CONTAINERS_NS` constant in `clean-up` commands

arnaldo2792 commented:
* Rebase base branch

arnaldo2792 commented:
@samuelkarp 

> If bootstrap containers can't be superpowered, we should enforce this assertion in the code of host-ctr
   If host containers and bootstrap containers are allowed to share the same name, we need to work out non-collision for the container IDs in containerd

* I updated to code to enforce that bootstrap containers can't be `superpowered`
* Bootstrap containers are now created in the `bootstrap` namespace to prevent collisions

@bcressey 

I marked all the systemd changes as resolved since I separated those changes in bottlerocket-os/bottlerocket#1423

arnaldo2792 commented:
* Removed `.expect(<>)` call and use `context` instead
* Updated documentation
* Removed repeated message in mode validation

arnaldo2792 commented:
* Rebase base branch

samuelkarp commented on README.md:
```diff
@@ -451,6 +451,46 @@ We use it for the control container because it needs to be available early to gi
 Be careful, and make sure you have a similar low-level use case before reaching for host containers.
 
+#### Bootstrap containers settings
+* `settings.bootstrap-containers.<name>.source`: the URI of the container
```
nit:
```suggestion
* `settings.bootstrap-containers.<name>.source`: the image for the container
```

Images are not specified as URIs since they lack a scheme.

arnaldo2792 replied:
Fixed

samuelkarp commented on README.md:
```diff
@@ -451,6 +451,46 @@ We use it for the control container because it needs to be available early to gi
 
+#### Bootstrap containers settings
+* `settings.bootstrap-containers.<name>.source`: the URI of the container
+* `settings.bootstrap-containers.<name>.mode`: the mode of the container, it could be one of `off`, `once` or `always`
```
What's the difference between `once` and `always`?  ~I assume that it's "first boot" vs. "every boot" but I don't think that's clear here.~

This is explained below, maybe add a pointer to it.

```suggestion
* `settings.bootstrap-containers.<name>.mode`: the mode of the container, it could be one of `off`, `once` or `always`.  See below for a description of modes.
```

arnaldo2792 replied:
Fixed

samuelkarp commented on README.md:
```diff
@@ -451,6 +451,46 @@ We use it for the control container because it needs to be available early to gi
+
+* `always`: with this setting, the container is executed on every boot.
+* `off`: the container won't run
+* `once`: with this setting, the container only runs on the first boot. Upon completion, the container marks itself as `off`.
```
```suggestion
* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is changed to `off`.
```

samuelkarp commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,722 @@
+* `off`: blocks the execution of the container
+* `once`: with this setting, the container only runs on the first boot. Upon completion,
+the container marks itself as `off`. If you need to execute the containers after the first
+boot, you can mark it as `once` and reboot.
```
```suggestion
* `once`: with this setting, the container only runs on the first boot where the container
is defined. Upon completion, the mode is changed to `off`. If you need to execute the
containers after the first boot, you can mark it as `once` and reboot.
```

samuelkarp commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,722 @@
+which can be used to fail the boot process if the underlying container task exits with a
+non-zero status code.
+
+Bootstrap containers are created in the `bootstrap` containerd's namespace to prevent
+collisions with other host containers, therefore you can have use the name for both container
+types.
```
I don't really think using a different namespace is the best approach here.  Namespaces extend to all parts of containerd's API, including image management.  Using a different namespace means that:

* someone wanting to reuse the same image for both bootstrap and regular host container with have to perform parts of the "pull" process twice (layers end up being shared in the default configuration, but you'll have to resolve the tag twice)
* removing images is no longer straightforward as they'll need to be removed from every namespace
* querying for container status is no longer straightforward as you'll need to query every namespace

Instead, it might be useful to introduce some separation between the name that we define in the Bottlerocket API and the underlying container ID; they don't need to exactly match.  For example, we could prefix host containers with `host.` and bootstrap containers with `boot.` and resolve the collision issue without introducing a new namespace (with the downside being losing 5 characters from the ID length limit of 76).

arnaldo2792 replied:
I thought on doing so but then I saw how tightly coupled is the `containerID` with other things like setting up the local mount filesystem (I had exactly the thing you asked for while I was implementing this change). If I prefix the container id with something within `host-ctr`, we need to extract what the "container name" is to setup the local persistent mount, otherwise existing customer images that already rely on `/.bottlerocket/host-containers/<container-name>` will be broken for the next release. 

These are the implementations that I tested (but didn't like them):

- pass `--container-id=<id> --container-type=<type>`, and let `host-ctr` figure out the correct prefix: I didn't like this idea because the caller is using a specific id and that id should be effectively used, but that is not the case because depending on the type we change the id and prefix it with something.

- pass `--container-id=<id> --container-type=<type>` where `--container-id` already contains the prefix: I like this idea better because now `host-ctr` effectively uses the passed container id. Still, within `host-ctr` there should be logic to get the correct container name to setup the local persistent mount.

samuelkarp replied:
> pass `--container-id=<id> --container-type=<type>`, and let host-ctr figure out the correct prefix

This is what I would recommend.

> I didn't like this idea because the caller is using a specific id and that id should be effectively used, but that is not the case because depending on the type we change the id and prefix it with something.

I don't really think we've made that guarantee.  That's the way host-ctr behaves today, but the way we implement host containers is sufficiently an implementation detail that I'm not afraid of changing it; we've even (long ago) considered whether we wanted to eventually move away from containerd and toward something like systemd-nspawn.  As long as the ID you use to interact with host-ctr remains the same, the ID that is used with containerd isn't a guaranteed interface.

If we _do_ want to keep the existing host containers with the same ID (for example because we have reason to believe that someone has taken a dependency on them), we could reserve the prefixing only for bootstrap containers.

jahkeup commented on README.md:
```diff
@@ -451,6 +451,46 @@ We use it for the control container because it needs to be available early to gi
+This type of container is very similar to normal host containers, they come with persistent storage and with optional user data.
+Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
+However, these containers have access to the underlying root filesystem on `/.bottlerocket/rootfs`.
+Bootstrap containers are started once the `configured` systemd's target is reached, so they are not executed in a deterministic order.
```
This might be slightly pedantic, but I think this more accurately clarifies why the dependency ordering can't provide deterministic ordering.

```suggestion
Bootstrap containers are set up to run after the systemd `configured.target` unit is active.
The containers' systemd unit depends on this target (and not on any of the bootstrap containers' peers) which means that bootstrap containers will not execute in a deterministic order.
```

arnaldo2792 replied:
Fixed

jahkeup commented on sources/api/README.md:
```diff
@@ -123,12 +123,13 @@ Some day we may need to make an earlier service (say, apiserver) configurable th
-This is a systemd target that depends on [settings-applier](#settings-applier) and represents the point at which the system is fully configured.
+### Default systemd target and init workflow
 
+Bottlerocket uses a custom default systemd target called `preconfigured`, which is used to initialize all the services used by the API system (i. e. [settings-applier](#settings-applier)).
```
I prefer that we use the full unit name when referring to systemd units in docs (and everywhere, really):

```suggestion
Bottlerocket configures a non-standard default systemd target, `preconfigured.target`, that coordinates initialization of the host and its services before starting the API system (i. e. [settings-applier](#settings-applier)).
```

arnaldo2792 replied:
I removed this change from the commits and I'll fix the documentation with #1448

samuelkarp commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,722 @@
+BEAR_DIR=/.bottlerocket/bootstrap-containers/bear
+VAR_DIR=/.bottlerocket/rootfs/var
+MY_DIR=$VAR_DIR/lib/my_directory
+cat $BEAR_DIR/user-data
```
I missed this, but the user data gets written to a file in the container directory?

arnaldo2792 replied:
Yes, as with [host-containers](https://github.com/bottlerocket-os/bottlerocket/blob/d76f82856329142e2dadb49da5eb44b45c93f35c/sources/api/host-containers/src/main.rs#L356)

samuelkarp commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,722 @@
+before services like the ecs-agent, kubelet and docker start.
+
+This type of containers are very similar to normal host containers, they come with
+persistent storage and with optional user data. Unlike normal host containers, bootstrap
```
nit:
```suggestion
Bootstrap containers are very similar to normal host containers; they come with
persistent storage and with optional user data. Unlike normal host containers, bootstrap
```

arnaldo2792 replied:
This documentation no longer exists in  the crate's README

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -38,6 +39,40 @@ import (
+	}
+
+	return errors.New("Invalid container type")
+}
```
In Go, it's typical to use a `bool` to indicate simple yes/no error states.  For example, when reading something that may or may not exist from a map, there's a `bool` returned and convention is to call it `ok`:

```go
val, ok := myMap[key]
```

For public methods, it's also [conventional](https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences) for comments documenting declarations to begin with the name of the identifier.

```suggestion
// IsValid returns true if the container type is valid
func (ct ContainerType) IsValid() bool {
	switch ct {
	case Host, Bootstrap:
		return true
	}
	return false
}
```

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -38,6 +39,40 @@ import (
 var ecrRegex = regexp.MustCompile(`(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?.*`)
 
+// Used to define valid container types
+type ContainerType string
```
nit: I'd move the type declaration and method bodies down either closer to where they're used or to the bottom of the file.

jahkeup commented on sources/api/README.md:
```diff
@@ -123,12 +123,13 @@ Some day we may need to make an earlier service (say, apiserver) configurable th
+### Default systemd target and init workflow
 
+Bottlerocket uses a custom default systemd target called `preconfigured`, which is used to initialize all the services used by the API system (i. e. [settings-applier](#settings-applier)).
+Once the target is reached, a systemd unit isolates the `configured` target, which represents the point at which the system is fully configured.
```
I'm not sure saying systemd "isolates" the unit in this context is clear enough. When systemd [`isolate`s to a given unit, systemd takes steps to take the system to a place where only the units that the _isolated unit_ defines (indirectly and directly) are active and inactive](https://www.freedesktop.org/software/systemd/man/systemctl.html#isolate%20UNIT) appropriately.  

```suggestion
Once `preconfigured.target` is reached, the system is isolated to the `configured.target` unit, which represents the point at which the system is fully configured.
```

arnaldo2792 replied:
I removed this from the commits and I'm going to fix the documentation with #1448

jahkeup commented on README.md:
```diff
@@ -451,6 +451,46 @@ We use it for the control container because it needs to be available early to gi
+* `settings.bootstrap-containers.<name>.essential`: whether or not the container should fail the boot process, default to `false`
+* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data
+
+Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.
```
I'd rather not call them host-containers if we're making a distinction between the two :)

```suggestion
Bootstrap containers, similar to host containers, are containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.
```

arnaldo2792 replied:
I wanted to make sure users understand that bootstrap containers have the same caveats as host containers, without repeating them, should I repeat the caveats?

jahkeup replied:
Seems worthwhile, maybe point to the docs saying "and the same caveats apply"

jahkeup commented on packages/os/bootstrap-containers-tmpfiles.conf:
```diff
@@ -0,0 +1,3 @@
+d /etc/bootstrap-containers 0755 root root -
```
Do we need to have world readable bits set on this directory? Can we instead restrict it to privileged users?

```suggestion
d /etc/bootstrap-containers 0750 root root -
```

bcressey replied:
I expect it'd be fine to change this, though we should change `host-containers-tmpfiles.conf` to match if we do.

arnaldo2792 replied:
Fixed

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -38,6 +39,40 @@ import (
+	}
+
+  return dir, err
+}
```
As above, I wouldn't use an `error` here.

```suggestion
// PersistentDir returns the persistent base directory for the container type
func (ct ContainerType) PersistentDir() (string, bool) {
	switch ct {
	case Host:
		return "host-containers", true
	case Bootstrap:
		return "bootstrap-containers", true
	}
	return "", false
}
```

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -3,6 +3,7 @@ package main
 import (
```
There's a bit of formatting funkyness going on with the file.  Can you make sure to run `gofmt`?

arnaldo2792 replied:
I ran go fmt, and I think the format is correct now :+1:

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -113,9 +149,15 @@ func App() *cli.App {
 				},
+				&cli.StringFlag{
+					Name:        "container-type",
+          Usage:       "specifies one of: [host, bootstrap]",
```
Funky formatting, please run `gofmt`.

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -383,8 +441,14 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
+  // Return error if container exists with non-zero status
+  if (code != 0) {
+    return errors.New("Container exited with non-zero status")
+  }
```
No parentheses are used around simple expressions.  Indentation is also off; should be fixed automatically by `gofmt`.

```suggestion
 	if code != 0 {
 		return errors.New("Container exited with non-zero status")
 	}
```

arnaldo2792 replied:
Fixed, `go fmt` fixed that for me :+1:

jahkeup commented on sources/api/README.md:
```diff
@@ -123,12 +123,13 @@ Some day we may need to make an earlier service (say, apiserver) configurable th
+Once the target is reached, a systemd unit isolates the `configured` target, which represents the point at which the system is fully configured.
 Applications can depend on this in their service definition.
 Services like Kubernetes and containerd depend on this.
+Upon completion, another systemd unit isolates the `multi-user` target, which is used to mark a successful boot.
```
```suggestion
After reaching `configured.target`, the system then isolates itself to `multi-user.target` which indicates that the system has fully booted.
At this point, when `multi-user.target` is activated, the system records itself as having completed a successful boot.
```

arnaldo2792 replied:
Oh, this is no longer true, we now mark the boot as successful on the `preconfigured` target, after `migrator.service` runs. Should I call that out here?

arnaldo2792 replied:
I removed this comment, and I'm going to fix this documentation with #1448

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -556,12 +620,25 @@ func withSuperpowered(superpowered bool) oci.SpecOpts {
 		oci.WithPrivileged,
 		oci.WithNewPrivileges,
+		oci.WithHostNamespace(runtimespec.PIDNamespace),
+		oci.WithParentCgroupDevices,
```
How come these are reordered?

arnaldo2792 replied:
Bad "revert", now they are in the order they were before.

jahkeup commented on sources/api/bootstrap-containers/README.md:
```diff
@@ -0,0 +1,75 @@
+types.
+
+## Examples
+Given a bootstrap container with the following configuration:
```
The example needs some introduction to help connect the dots on some of the bits in the bootstrap script.

```suggestion
Given a bootstrap container called `bear` with the following configuration:
```

arnaldo2792 replied:
Fixed

samuelkarp commented on sources/models/src/aws-k8s-1.15/mod.rs:
```diff
@@ -16,6 +16,7 @@ struct Settings {
     kubernetes: KubernetesSettings,
     updates: UpdatesSettings,
     host_containers: HashMap<Identifier, ContainerImage>,
+    bootstrap_containers: HashMap<Identifier, BootstrapContainer>,
```
If we're using the map key as the container ID, we should probably create a new type that validates the length.  containerd enforces a 76-character maximum on the container ID.

arnaldo2792 replied:
I'll make this update just for bootstrap containers, ~~I don't want to break existing customers for normal host containers.~~

Actually, it will be safe to validate all containers since all variants use containerd

arnaldo2792 replied:
I added a validation to `Identifier` to check for the length

jahkeup commented on sources/api/bootstrap-containers/README.md:
````diff
@@ -0,0 +1,75 @@
+And `bootstrap-script` as:
+
+```sh
+#!/bin/sh
+BEAR_DIR=/.bottlerocket/bootstrap-containers/bear
+VAR_DIR=/.bottlerocket/rootfs/var
+MY_DIR=$VAR_DIR/lib/my_directory
+cat $BEAR_DIR/user-data
+mkdir $MY_DIR
+chmod -R o+r $MY_DIR
+chown -R 1000:1000 $MY_DIR
+sleep 120
````
I suggest we take advantage of [the `$prefix/current` directory which has the user-data](https://github.com/bottlerocket-os/bottlerocket/pull/1416) in this example. We should encourage users to develop their own containers against that path instead of hard coding expected `bootstrap-containers` names within scripts:

```suggestion
#!/usr/bin/env sh
# We'll read some data to be written out from given user-data.
USER_DATA_DIR=/.bottlerocket/bootstrap-containers/current
# This is the in-container view of where the host's `/var` can be accessed.
HOST_VAR_DIR=/.bottlerocket/rootfs/var
# The directory that'll be created by this bootstrap container
MY_HOST_DIR=$VAR_DIR/lib/my_directory

# Create it!
mkdir -p "$MY_HOST_DIR"

# Write the user-data to stdout (to the journal) and to our new path:
tee /dev/stdout "$MY_HOST_DIR/bear.txt" < "$USER_DATA_DIR/user-data" 

# The bootstrap container can set the permissions which are seen by the host:
chmod -R o+r "$MY_HOST_DIR"
chown -R 1000:1000 "$MY_HOST_DIR"

# Bootstrap containers *must* finish before boot continues.
# 
# With this, the boot process will be delayed 120 seconds. You can check the
# status of `preconfigured.target` and `bootstrap-containers@bear` to see
# that this sleep kept the system from starting up the apiserver.
#
# From the admin container:
#
# systemctl status preconfigured.target bootstrap-containers@bear
sleep 120
```

arnaldo2792 replied:
Fixed

arnaldo2792 commented:
* Updated the `bootstrap-containers@service` template to prevent manual interactions with the units through `systemctl`

jahkeup commented on sources/api/bootstrap-containers/README.md:
````diff
@@ -0,0 +1,75 @@
+sleep 120
+```
+
+You should see a new directory under `/var/lib` called `my_directory`, and
+the following command should show something similar to `ʕ·͡ᴥ·ʔ`:
````
```suggestion
You should see a new directory under `/var/lib` called `my_directory`, a file in that directory called `bear.txt`, and the following command should show `ʕ·͡ᴥ·ʔ`
in the bootstrap-container's logs:
```

arnaldo2792 replied:
Fixed

jahkeup reviewed (commented):
Still reading through - but wanted to share some feedback.

arnaldo2792 commented:
I changed the base back to develop since bottlerocket-os/bottlerocket#1423 was merged

bcressey commented on packages/os/bootstrap-containers@.service:
```diff
@@ -0,0 +1,23 @@
+[Unit]
+Description=bootstrap container %i
+Before=configured.target
+After=host-containerd.service
```
Should also have a dependency:
```suggestion
After=host-containerd.service
Wants=host-containerd.service
```

arnaldo2792 replied:
Done

bcressey commented on packages/os/bootstrap-containers@.service:
```diff
@@ -0,0 +1,23 @@
+[Service]
+Type=oneshot
+EnvironmentFile=/etc/bootstrap-containers/%i.env
+Environment=LOCAL_DIR=/local
```
This is a remnant from when we ran `mkdir` here. We should also remove it from `host-containers@.service`.
```suggestion
```

arnaldo2792 replied:
Done

bcressey commented on sources/models/src/aws-ecs-1/mod.rs:
```diff
@@ -15,6 +15,7 @@ struct Settings {
     motd: String,
     updates: UpdatesSettings,
     host_containers: HashMap<Identifier, ContainerImage>,
+    bootstrap_containers: HashMap<Identifier, BootstrapContainer>,
```
We should refactor the host container type to be `HostContainer` rather than `ContainerImage`.

arnaldo2792 replied:
Done

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -237,3 +237,13 @@ struct Metadata {
+#[model]
+struct BootstrapContainer {
+    source: Url,
+    mode: ValidContainerMode,
```
Let's scope this type to bootstrap containers:
```suggestion
    mode: ValidBootstrapContainerMode,
```

arnaldo2792 replied:
Done

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -137,6 +140,18 @@ macro_rules! string_impls_for {
             }
         }
+
+        impl PartialEq<str> for $for {
+            fn eq(&self, other: &str) -> bool {
+                &self.inner == other
+            }
+        }
+
+        impl PartialEq<String> for $for {
+            fn eq(&self, other: &String) -> bool {
+                &self.inner == other
+            }
+        }
```
Why did we need these impls?

arnaldo2792 replied:
So we can compare string-like types with other strings, i.e.:

```rust
let a_string = "a_string".to_string();
let mode = ValidContainerMode{ inner: a_string };

assert!(mode == a_string);
```

arnaldo2792 committed:
models: change ContainerImage to HostContainer

This commit updates the model name used to configure a host container
from `ContainerImage` to `HostContainer`

arnaldo2792 committed:
host-containers: change permissions in `/etc/host-containers`

This commit updates the permissions in `/etc/host-containers` from `755`
to `750`

arnaldo2792 commented:
- Use prefixes instead of namespaces in host-ctr
- Validate `Identifier` length
- Change `ContainerImage` to `HostContainer`
- Fix documentation with provided  feedback
- Change permissions on both `/etc/host-containers` and `/etc/bootstrap-containers` to `750`
- Use bools instead of errors in `ContainerType` validations

tjkirch commented on sources/models/src/lib.rs:
```diff
@@ -241,3 +241,13 @@ struct Metadata {
+///// Bootstrap Containers
+
+#[model]
+struct BootstrapContainer {
+    source: Url,
+    mode: ValidBootstrapContainerMode,
+    user_data: ValidBase64,
+    essential: bool,
+}
```
Without a `superpowered` flag like host containers, how do users control the level of permissions they need?  One use case we know of is using `apiclient apply` during early startup to apply settings from a file or URL, if it's too big for user data or they want to share settings.  They don't need access to modify disks, or anything else superpowered.  Other use cases, like setting up disks, need a lot more access.

tjkirch replied:
Some further discussion led to the decision that bootstrap containers will get the minimal extra permissions they need to serve their purpose (ability to set up the filesystem, disks, etc.) but shouldn't have a superpowered flag because it could be abused, even unintentionally, to break the system in obscure ways.

bcressey commented on README.md:
```diff
@@ -478,6 +478,47 @@ We use it for the control container because it needs to be available early to gi
+
+* `always`: with this setting, the container is executed on every boot.
+* `off`: the container won't run
+* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is chaged to `off`.
```
```suggestion
* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is changed to `off`.
```

arnaldo2792 replied:
Fixed

bcressey commented on README.md:
```diff
@@ -478,6 +478,47 @@ We use it for the control container because it needs to be available early to gi
+* `always`: with this setting, the container is executed on every boot.
+* `off`: the container won't run
+* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is chaged to `off`.
+If you need to execute the containers after the first boot, you can mark it as `once` and reboot.
```
nit: not sure we need this note, it seems like the result of this might not be very well-defined.

bcressey commented on README.md:
```diff
@@ -478,6 +478,47 @@ We use it for the control container because it needs to be available early to gi
+[settings.bootstrap-containers.bootstrap]
+source = "MY-CONTAINER-URI"
+mode = "once"
+essential = false
```
Should we suggest `essential = true` as the default? It's more in keeping with the behavior of the overall system.

arnaldo2792 replied:
Done

bcressey commented on README.md:
```diff
@@ -478,6 +478,47 @@ We use it for the control container because it needs to be available early to gi
+
+Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.
+
+This type of container is very similar to normal host containers, they come with persistent storage and with optional user data.
```
```suggestion
These container are very similar to normal host containers.
They come with persistent storage and with optional user data.
```

arnaldo2792 replied:
Fixed

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,708 @@
+* `off`: blocks the execution of the container
+* `once`: with this setting, the container only runs on the first boot. Upon completion,
+the container marks itself as `off`. If you need to execute the containers after the first
+boot, you can mark it as `once` and reboot.
```
The examples might be OK, though this is a weird place for them. I'd prefer that the content read more like what we have in `sources/api/host-containers/README.md` though, and not just repeat from the main README. The duplication makes me worry about keeping them in sync.

arnaldo2792 replied:
I updated the documentation

bcressey commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -383,8 +452,14 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
+
 	log.G(ctrCtx).WithField("code", code).Info("container task exited")
 
+	// Return error if container exists with non-zero status
+	if code != 0 {
+		return errors.New("Container exited with non-zero status")
+	}
+
```
Have you tested that ordinary host containers are still properly restarted after the change?

It would be good to include the name of the container that failed here for additional context.

arnaldo2792 replied:
If by restarted you mean `kill -9 <host-container-pid>`, and wait for the container to come back up, then yes, that still works:

```sh
ps aux | grep host-ctr
root         872  0.1  0.9 1245748 35696 ?       Ssl  00:20   0:00 /usr/bin/host-ctr run --container-id=control

sudo kill -9 872

systemctl status host-containers@control.service

host-containers@control.service - Host container: control
     Loaded: loaded (/aarch64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/host-containers@.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: signal) since Wed 2021-04-07 00:22:21 UTC; 21s ago
    Process: 872 ExecStart=/usr/bin/host-ctr run --container-id=control --source=${CTR_SOURCE} --superpowered=${CTR_SUPERPOWERED} (code=killed, signal=KILL)

# waited for some time
systemctl status host-containers@control.service

host-containers@control.service - Host container: control
     Loaded: loaded (/aarch64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/host-containers@.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2021-04-07 00:23:06 UTC; 5s ago
```

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,708 @@
+        }
+    }
+
+    match subcommand.as_deref() {
+        // Default subcommand is 'create-containers'
+        None | Some("create-containers") => Ok((global_args, Subcommand::CreateContainers {})),
+        Some("mark-bootstrap") => Ok((global_args, parse_mark_bootstrap_args(subcommand_args)?)),
+        Some(x) => {
+            return error::Usage {
+                message: format!("Unknown subcommand '{}'", x),
+            }
+            .fail()
+        }
+    }
+}
```
I'd prefer to just error out on an unknown subcommand.

arnaldo2792 replied:
Done

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,708 @@
+
+    info!("Handling bootstrap container '{}'", name);
+
+    // Get basic settings, as retrieved from API.
+    let source = container_details
+        .source
+        .as_ref()
+        .context(error::MissingField {
+            name,
+            field: "source",
+        })?;
+
+    let mode = container_details
+        .mode
+        .as_ref()
+        .context(error::MissingField {
+            name,
+            field: "mode",
+        })?;
```
This creates the same concern as #1385 for bootstrap containers.

I would say:
* if no mode, default to "off"
* if not essential, default to "false" (as you've done)

arnaldo2792 replied:
I'm still working on fixing this

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,708 @@
+    }
+
+    fn enable(&self) -> Result<()> {
+        // Only enable the unit, since it will be started once systemd reaches the `configured`
+        // target
```
nit: the unit is started after systemd reaches the `preconfigured` target

arnaldo2792 replied:
Fixed

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,708 @@
+        #[snafu(display("Failed to manage {} of {} bootstrap containers", failed, tried))]
+        ManageContainersFailed { failed: usize, tried: usize },
+
+        #[snafu(display("Host containers '{}' missing field '{}'", name, field))]
```
```suggestion
        #[snafu(display("Bootstrap containers '{}' missing field '{}'", name, field))]
```

arnaldo2792 replied:
Fixed

arnaldo2792 commented:
* Fixed documentation feedback
* Show status code in `commad` function executed by bootstrap containers

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,730 @@
+    let mode = args.mode.as_ref();
+    info!("Mode for '{}' is '{}'", container_id, mode);
+
+    // When 'mode' is 'once', the containers is marked as 'off' once it
+    // finishes. This guarantees that the the container is only started in
+    // the first boot.
+    if mode != "always" {
+        let formatted = format!("settings.bootstrap-containers.{}.mode", container_id);
+        let key =
+            Key::new(KeyType::Data, &formatted).context(error::KeyFormat { key: formatted })?;
+        let value = serialize_scalar(&"off".to_string()).context(error::Serialize)?;
+
+        let mut map = HashMap::new();
+        map.insert(key, value);
+        let settings: model::Settings = datastore::deserialization::from_map(&map)
+            .context(error::SettingsDeserialize { settings: map })?;
+
+        info!("Turning off container '{}'", container_id);
+        apiclient::set::set(socket_path, &settings)
+            .await
+            .context(error::Set)?;
+    }
```
It might be worth adding a comment that this causes another, possibly concurrent invocation of `bootstrap-containers` after the API setting finalizes, which is probably safe but could be fun to debug if not.

arnaldo2792 replied:
Done

arnaldo2792 commented:
* Fixed typo in documentation
* Removed repeated documentation in crate
* Error out when no command is passed to `bootstrap-containers`
* Suggested `essential=true` in documentation
* Added container id in non-zero status error message in `host-ctr`
* Removed status code from `bootstrap-containers` `command` function
* Removed `Valid` from `BootstrapContainerMode`

arnaldo2792 commented:
* Use `off` as default for `mode`
* Removed a note from the documentation
* Removed `--now` from disable command and added a note explaining why

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -0,0 +1,715 @@
+    }
+}
+
+/// Parses arguments for the 'mark-boostrap' subcommand
```
nit:
```suggestion
/// Parses arguments for the 'mark-bootstrap' subcommand
```

arnaldo2792 commented:
* Rebase upstream
* Fixed typo

arnaldo2792 commented:
* Updated the commit message that mentioned using namespaces instead of the prefix in `host-ctr`

arnaldo2792 commented:
* Fixed documentation in bootstrap containers

jahkeup commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -154,7 +162,59 @@ func App() *cli.App {
+	return false
+}
+
+// PersistentDir returns the persistent base directory for the container type
+func (ct ContainerType) PersistentDir() (string, bool) {
+	switch ct {
+	case Host:
+		return "host-containers", true
+	case Bootstrap:
+		return "bootstrap-containers", true
+	}
+
+	return "", false
+}
```
The call sites for these don't seem to use the third state, and is not idiomatic in this case (return the zero value). Can we drop this from the signature please?

```suggestion
// PersistentDir returns the persistent base directory for the container type
func (ct ContainerType) PersistentDir() string {
	switch ct {
	case Host:
		return "host-containers"
	case Bootstrap:
		return "bootstrap-containers"
	}

	return ""
}
```

jahkeup commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -154,7 +162,59 @@ func App() *cli.App {
+	return "", false
+}
+
+// Prefix returns the prefix for the container type
+func (ct ContainerType) Prefix() (string, bool) {
+	switch ct {
+	case Bootstrap:
+		return "boot.", true
+	case Host:
+		return "", true
+	}
+
+	return "", false
+}
```
The call sites for these don't seem to use the third state, and is not idiomatic in this case (return the zero value). Can we drop this from the signature please?

```suggestion
// Prefix returns the prefix for the container type
func (ct ContainerType) Prefix() string {
	switch ct {
	case Bootstrap:
		return "boot."
	case Host:
		return ""
	}

	return ""
}
```

jahkeup reviewed (changes requested):
There's a couple small changes I'd like to see in `host-ctr` before merge.

arnaldo2792 commented:
* Rebase upstream

jahkeup commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -154,7 +162,59 @@ func App() *cli.App {
 }
 
-func runCtr(containerdSocket string, namespace string, containerID string, source string, superpowered bool) error {
+// Used to define valid container types
+type ContainerType string
+
+const (
+	Host      ContainerType = "host"
+	Bootstrap ContainerType = "bootstrap"
+)
```
GitHub lost my comment.

Please update this to keep the types private.

```suggestion
// Used to define valid container types
type containerType string

const (
	containerTypeHost      containerType = "host"
	containerTypeBootstrap containerType = "bootstrap"
)
```

arnaldo2792 commented:
* Refactored `ContainerType`, `Host` and `Bootstrap` to `containerType`, `host` and `bootstrap`. I had to use `cType` as variables' names to prevent collisions

arnaldo2792 commented:
* I rebased upstream, and had to update some rust dependencies as part of the rebase
* Removed `user-data` base64 string from error message (shouldn't be part of the rebased change, but I forge to not merge it)
